pkg/metricsadapter/provider/resourcemetrics.go

```diff
@@ -0,0 +1,463 @@
+}
+
+// queryNodeMetricsByName queries metrics by node name from target clusters
+func (r *ResourceMetricsProvider) queryNodeMetricsByName(name string) ([]metrics.NodeMetrics, error) {
+	clusters, err := r.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+
+	// step 1. Find out the target clusters with lister cache
+	var targetClusters []string
+	nodeGVR := corev1.SchemeGroupVersion.WithResource("nodes")
+	for _, cluster := range clusters {
+		sci := r.informerManager.GetSingleClusterManager(cluster.Name)
+		if sci == nil {
+			klog.Errorf("Failed to get cluster(%s) manager", cluster.Name)
+			continue
+		}
+		_, err := sci.Lister(nodeGVR).Get(name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				klog.Errorf("Failed to get node(%s) in cluster(%s): %v", name, cluster.Name, err)
+			}
+			continue
+		}
+		targetClusters = append(targetClusters, cluster.Name)
+	}
+
+	// step 2. Query metrics from the filtered target clusters
+	var nodeMetrics []metrics.NodeMetrics
+	for _, clusterName := range targetClusters {
+		sci := r.informerManager.GetSingleClusterManager(clusterName)
+		if sci == nil {
+			klog.Errorf("Failed to get cluster(%s) manager", clusterName)
+			continue
+		}
+
+		metricsV1betaGVR := metricsv1beta1.SchemeGroupVersion.WithResource("nodes")
+		metrics, err := sci.GetClient().Resource(metricsV1betaGVR).Get(context.Background(), name, metav1.GetOptions{})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				klog.Errorf("Failed to get node(%s) metrics in cluster(%s): %v", name, clusterName, err)
+			}
+			continue
+		}
+		internalMetrics, err := metricsConvertV1beta1NodeToInternalNode(*metrics)
+		if err != nil {
+			return nil, err
+		}
+
+		nodeMetrics = append(nodeMetrics, internalMetrics...)
+	}
+
+	return nodeMetrics, nil
+}
+
+// queryNodeMetricsBySelector queries metrics by node selector from target clusters
+func (r *ResourceMetricsProvider) queryNodeMetricsBySelector(selector string) ([]metrics.NodeMetrics, error) {
+	labelSelector, err := labels.Parse(selector)
+	if err != nil {
+		klog.Errorf("Failed to parse label selector: %v", err)
+		return nil, err
+	}
+
+	clusters, err := r.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+
+	// step 1. Find out the target clusters with lister cache
+	var targetClusters []string
+	nodesGVR := corev1.SchemeGroupVersion.WithResource("nodes")
+	for _, cluster := range clusters {
+		sci := r.informerManager.GetSingleClusterManager(cluster.Name)
+		if sci == nil {
+			klog.Errorf("Failed to get cluster(%s) manager", cluster.Name)
+			continue
+		}
+		nodes, err := sci.Lister(nodesGVR).List(labelSelector)
+		if err != nil {
+			klog.Errorf("Failed to list nodes in cluster(%s): %v", cluster.Name, err)
+			continue
+		}
+		if len(nodes) != 0 {
+			targetClusters = append(targetClusters, cluster.Name)
```

Comment:
No, it's not an error, just return an empty list, and the metrics frame work will handle it for us.
If the list is empty, it will return `resource not found`.

Reply:
> No, it's not an error, just return an empty list, and the metrics frame work will handle it for us.

ok， that's fine.
>If the list is empty, it will return resource not found.
If the node not empty, just not meet selector, it will not return **resource not found**. also a empty。
see
`
// ListAll calls appendFn with each value retrieved from store which matches the selector.
func ListAll(store Store, selector labels.Selector, appendFn AppendFunc) error {
	selectAll := selector.Empty()
	for _, m := range store.List() {
		if selectAll {
			// Avoid computing labels of the objects to speed up common flows
			// of listing all objects.
			appendFn(m)
			continue
		}
		metadata, err := meta.Accessor(m)
		if err != nil {
			return err
		}
		if selector.Matches(labels.Set(metadata.GetLabels())) {
			appendFn(m)
		}
	}
	return nil
}
`